### PR TITLE
Add BEEFY capabilities to Westend and Kusama

### PR DESCRIPTION
**Pull request by acatangiu**
This enables BEEFY capabilities on Westend and Kusama.

Governance/sudo call is later required to enable/start consensus.

Part of https://github.com/paritytech/parity-bridges-common/issues/1634

### Notes:

Validators can still opt-out of running BEEFY using `--no-beefy` flag, but this should be used in case of stability issues only. Ideally validators should run BEEFY consensus.

Westend and Kusama validators (or at least archive nodes), should also run with `--enable-offchain-indexing=true` so that MMR proof generation is available for submitting to light-clients.

cumulus companion: https://github.com/paritytech/cumulus/pull/3021

### CODE REVIEW

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot/-/jobs/3348601

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot/-/jobs/3348956

**Review by andresilva — CHANGES_REQUESTED** [review]
I haven't reviewed the entire PR yet, was mainly looking at session key handling (since it's something that could break the existing runtime).

**Comment by andresilva** [comment]
Also regarding session keys, when this gets deployed we'll have to ask validators to rotate their session keys so that BEEFY keys are generated and set on-chain. We'll need to monitor this before triggering the activation.

**Comment by Polkadot-Forum** [comment]
This pull request has been mentioned on **Polkadot Forum**. There might be relevant details there:

https://forum.polkadot.network/t/polkadot-kusama-bridge/2971/4

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot/-/jobs/3360702

File: runtime/westend/src/lib.rs [hunk 1]
```diff
@@ -314,6 +389,7 @@ impl_opaque_keys! {
 		pub para_validator: Initializer,
 		pub para_assignment: ParaSessionInfo,
 		pub authority_discovery: AuthorityDiscovery,
+		pub beefy: Beefy,
```
[the remark this answers is not in the document]
Comment (acatangiu) [reply runtime/westend/src/lib.rs:403]: done

**Review by acatangiu — COMMENTED** [review]
@ggwpez or @kianenigma this is my first runtime migration, how do I properly test it? 😄

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot/-/jobs/3382071

**Comment by acatangiu** [comment]
> @ggwpez or @kianenigma this is my first runtime migration, how do I properly test it? 😄

I tested this locally and it (now) works fine. @andresilva PTAL

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot/-/jobs/3392438

File: runtime/kusama/src/lib.rs [hunk 2]
```diff
@@ -1521,7 +1642,7 @@ impl Get<Perbill> for NominationPoolsMigrationV4OldPallet {
 ///
 /// This contains the combined migrations of the last 10 releases. It allows to skip runtime
 /// upgrades in case governance decides to do so. THE ORDER IS IMPORTANT.
-pub type Migrations = (migrations::Unreleased,);
+pub type Migrations = (migrations::Unreleased, migrations::UpgradeSessionKeys);
```
Comment (andresilva) [inline runtime/kusama/src/lib.rs:1645]: Not sure whether this should be added here or directly to `migrations::Unreleased`, hopefully someone else will comment on this.

**Review by andresilva — APPROVED** [review]
lgtm. just minor nits

**Comment by acatangiu** [comment]
bot rebase

File: runtime/kusama/src/lib.rs (hunk 2 above)
Comment (bkchr) [reply runtime/kusama/src/lib.rs:1653]: Yeah, this needs to be added to the `Unreleased` type below.

File: runtime/kusama/src/lib.rs [hunk 3]
```diff
@@ -1607,6 +1730,16 @@ pub mod migrations {
...
 
 	);
+
+	// Upgrade Session keys to include BEEFY key.
+	// When this is removed, should also remove `OldSessionKeys`.
+	pub struct UpgradeSessionKeys;
+	impl frame_support::traits::OnRuntimeUpgrade for UpgradeSessionKeys {
+		fn on_runtime_upgrade() -> Weight {
+			Session::upgrade_keys::<OldSessionKeys, _>(transform_session_keys);
+			Perbill::from_percent(50) * BlockWeights::get().max_block
```
Comment (bkchr) [inline runtime/kusama/src/lib.rs:1740]:
```suggestion
			BlockWeights::get().max_block
```

50% sounds rather random. Just let's take all the weight.

File: runtime/westend/src/lib.rs [hunk 4]
```diff
@@ -1271,7 +1392,7 @@ impl Get<Perbill> for NominationPoolsMigrationV4OldPallet {
 ///
 /// This contains the combined migrations of the last 10 releases. It allows to skip runtime
 /// upgrades in case governance decides to do so. THE ORDER IS IMPORTANT.
-pub type Migrations = migrations::Unreleased;
+pub type Migrations = (migrations::Unreleased, migrations::UpgradeSessionKeys);
```
[the remark this answers is not in the document]
Comment (bkchr) [reply runtime/westend/src/lib.rs:1403]: Needs to be done here as well.

File: runtime/westend/src/lib.rs [hunk 5]
```diff
@@ -1292,6 +1415,16 @@ pub mod migrations {
...
 		parachains_configuration::migration::v8::MigrateToV8<Runtime>,
 	);
+
+	// Upgrade Session keys to include BEEFY key.
+	// When this is removed, should also remove `OldSessionKeys`.
+	pub struct UpgradeSessionKeys;
+	impl frame_support::traits::OnRuntimeUpgrade for UpgradeSessionKeys {
+		fn on_runtime_upgrade() -> Weight {
+			Session::upgrade_keys::<OldSessionKeys, _>(transform_session_keys);
+			Perbill::from_percent(50) * BlockWeights::get().max_block
```
Comment (bkchr) [inline runtime/westend/src/lib.rs:1425]:
```suggestion
			BlockWeights::get().max_block
```

File: runtime/westend/src/lib.rs [hunk 6]
```diff
@@ -1292,6 +1415,16 @@ pub mod migrations {
 		parachains_scheduler::migration::v1::MigrateToV1<Runtime>,
 		parachains_configuration::migration::v8::MigrateToV8<Runtime>,
 	);
+
+	// Upgrade Session keys to include BEEFY key.
+	// When this is removed, should also remove `OldSessionKeys`.
```
Comment (bkchr) [inline runtime/westend/src/lib.rs:1420]:
```suggestion
	/// Upgrade Session keys to include BEEFY key.
	/// When this is removed, should also remove `OldSessionKeys`.
```

**Review by bkchr — APPROVED** [review]
Only nitpicks. Look good otherwise :)

**Comment by acatangiu** [comment]
bot merge

**Pull request merged**